src/Libraries/RevitNodes/Application/Document.cs

```diff
@@ -93,6 +108,173 @@ internal ModelPath WorksharingModelPath
+                .OfClass(typeof(Autodesk.Revit.DB.Material))
+                .ToElementIds()
+                .ToList();
+
+            if (materials == null)
+                return new List<ElementId>();
+
+            List<Autodesk.Revit.DB.Element> elementTypes = new FilteredElementCollector(this.InternalDocument)
+                .WhereElementIsElementType()
+                .ToElements()
+                .ToList();
+
+            if (elementTypes == null)
+                return new List<ElementId>();
+
+            var nonPurgedElements = new HashSet<ElementId>();
+
+            for (int i = 0; i < elementTypes.Count; i++)
+            {
+                nonPurgedElements.UnionWith(GetNonPurgableMaterialIds(elementTypes[i]));
+            }
+
+            List<ElementId> purgableMaterials = materials.Except(nonPurgedElements).ToList();
+            if (purgableMaterials.Count > 0)
+                this.InternalDocument.Delete(purgableMaterials);
+
+            List<ElementId> purgedMaterials = new List<ElementId>();
+            purgedMaterials.AddRange(PurgeMaterialAssets(nonPurgedElements.ToList()));
+            purgedMaterials.AddRange(purgableMaterials);
+
+            return purgedMaterials;
+        }
+
+        private static HashSet<ElementId> GetNonPurgableMaterialIds(Autodesk.Revit.DB.Element type)
+        {
+            HostObjAttributes hostObj = type as HostObjAttributes;
+            HashSet<ElementId> usedMaterialIds = new HashSet<ElementId>();
+            if (hostObj != null)
+            {
+                var compundStructure = hostObj.GetCompoundStructure();
+                if (compundStructure != null)
+                {
+                    int layerCount = compundStructure.LayerCount;
+                    for (int j = 0; j < layerCount; j++)
+                    {
+                        usedMaterialIds.Add(compundStructure.GetMaterialId(j));
+                    }
+                }
+            }
+
+            List<ElementId> elementMaterialIds = type.GetMaterialIds(false).ToList();
+            if (elementMaterialIds.Any())
+            {
+                usedMaterialIds.UnionWith(elementMaterialIds);
+            }
+            return usedMaterialIds;
+        }
+
+        private List<Autodesk.Revit.DB.ElementId> PurgeMaterialAssets(List<Autodesk.Revit.DB.ElementId> elementIds)
+        {
+            List<ElementId> appearanceAssetIds = new FilteredElementCollector(this.InternalDocument)
+                .OfClass(typeof(AppearanceAssetElement))
+                .ToElementIds()
+                .ToList();
+
+            List<ElementId> propertySet = new FilteredElementCollector(this.InternalDocument)
+                .OfClass(typeof(PropertySetElement))
+                .ToElementIds()
+                .ToList();
+
+            int elementIdsCount = elementIds.Count();
+            for (int i = 0; i < elementIdsCount; i++)
```

Comment:
need a comment here explaining what it does, never came across `PropertySet` in RevitAPI before, so others likely haven't either